node/pkg/watchers/aptos/watcher.go

```diff
@@ -282,6 +501,87 @@ func (e *Watcher) retrievePayload(s string) ([]byte, error) {
 	return body, err
 }
 
+//nolint:noctx // TODO: this function should use a context.
+func (e *Watcher) queryIndexer(query string, headers map[string]string) ([]byte, error) {
+	// Create GraphQL request body
+	requestBody := map[string]interface{}{
+		"query": query,
+	}
+	jsonBody, err := json.Marshal(requestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", e.aptosIndexerRPC, bytes.NewBuffer(jsonBody))
+	if err != nil {
+		return nil, err
+	}
+
+	// Set content type for GraphQL
+	req.Header.Set("Content-Type", "application/json")
+
+	// Add additional headers if provided
+	for key, value := range headers {
+		req.Header.Set(key, value)
+	}
+
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	body, err := common.SafeRead(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	return body, err
+}
+
+// processTransactionVersion fetches a transaction by version and extracts WormholeMessage events
+func (e *Watcher) processTransactionVersion(
+	logger *zap.Logger,
+	versionNum uint64,
+	sequenceForObserve uint64,
+	isReobservation bool,
+) error {
+	// Fetch transaction details for this version
+	var txEndpoint = fmt.Sprintf(`%s/v1/transactions/by_version/%d`, e.aptosRPC, versionNum)
+	txData, err := e.retrievePayload(txEndpoint)
+	if err != nil {
+		logger.Error("retrievePayload", zap.Error(err))
+		p2p.DefaultRegistry.AddErrorCount(e.chainID, 1)
+		return err
+	}
+
+	if !gjson.Valid(string(txData)) {
+		logger.Error("Invalid JSON in transaction response", zap.String("data", string(txData)))
+		p2p.DefaultRegistry.AddErrorCount(e.chainID, 1)
+		return fmt.Errorf("invalid JSON in transaction response")
+	}
+
+	// Parse the transaction data
+	txResult := gjson.ParseBytes(txData)
+
+	// Look for WormholeMessage in the events array
+	whEventType := fmt.Sprintf("%s::state::WormholeMessage", e.aptosAccount)
+	events := txResult.Get("events")
+	if events.Exists() {
+		for _, event := range events.Array() {
```

Comment:
Should this have a length check? The loop will be skipped silently if the array is empty.